File: Java/LaptopAsus.java
public class LaptopAsus {
    private final String merk = "ASUS";
    private String model;
    private String nomorSeri;
    private int kapasitasBaterai; // 0 - 100
    private boolean statusMenyala;
    private int jumlahRAM;

    public LaptopAsus(String model, String nomorSeri, int jumlahRAM) {
        this.model = model;
        this.nomorSeri = nomorSeri;
        this.jumlahRAM = jumlahRAM;
        this.kapasitasBaterai = 100; // default penuh saat dibuat
        this.statusMenyala = false;
    }

    // Invarian 1: kapasitas baterai selalu 0-100
    public void cashBaterai(int jumlah) {
        if (jumlah < 0) {
            System.out.println("Jumlah cas tidak valid.");
            return;
        }
        kapasitasBaterai = Math.min(100, kapasitasBaterai + jumlah);
    }

    public void pakaiBaterai(int jumlah) {
        if (jumlah < 0) {
            System.out.println("Jumlah pemakaian tidak valid.");
            return;
        }
        kapasitasBaterai = Math.max(0, kapasitasBaterai - jumlah);
        if (kapasitasBaterai == 0) {
            statusMenyala = false; // otomatis mati jika baterai habis
        }
    }

    // Invarian 2: tidak bisa menyala jika baterai 0
    public void nyalakan() {
        if (kapasitasBaterai <= 0) {
            System.out.println("Gagal menyalakan: baterai habis!");
            return;
        }
        statusMenyala = true;
        System.out.println(model + " berhasil dinyalakan.");
    }

    public void matikan() {
        statusMenyala = false;
        System.out.println(model + " dimatikan.");
    }

    public void getInfo() {
        System.out.println("Merk        : " + merk);
        System.out.println("Model       : " + model);
        System.out.println("Nomor Seri  : " + nomorSeri);
        System.out.println("RAM         : " + jumlahRAM + " GB");
        System.out.println("Baterai     : " + kapasitasBaterai + "%");
        System.out.println("Status      : " + (statusMenyala ? "Menyala" : "Mati"));
    }

    public static void main(String[] args) {
        LaptopAsus laptop = new LaptopAsus("ROG Strix G16", "ASUS-2024-001", 16);
        laptop.getInfo();

        // Operasi sah
        laptop.nyalakan();
        laptop.pakaiBaterai(30);
        laptop.getInfo();

        // Operasi tidak sah 1: cas melebihi 100
        laptop.cashBaterai(200);
        laptop.getInfo(); // tetap maksimal 100

        // Operasi tidak sah 2: nyalakan saat baterai habis
        laptop.pakaiBaterai(100); // baterai jadi 0, otomatis mati
        laptop.nyalakan(); // harus ditolak
        laptop.getInfo();
    }
}

File: Pertemuan 1/php/LaptopAsus.php
<?php
class LaptopAsus {
    private string $merk = "ASUS";
    private string $model;
    private string $nomorSeri;
    private int $kapasitasBaterai; // 0 - 100
    private bool $statusMenyala;
    private int $jumlahRAM;

    public function __construct(string $model, string $nomorSeri, int $jumlahRAM) {
        $this->model = $model;
        $this->nomorSeri = $nomorSeri;
        $this->jumlahRAM = $jumlahRAM;
        $this->kapasitasBaterai = 100;
        $this->statusMenyala = false;
    }

    // Invarian 1: kapasitas baterai selalu 0-100
    public function cashBaterai(int $jumlah): void {
        if ($jumlah < 0) {
            echo "Jumlah cas tidak valid.\n";
            return;
        }
        $this->kapasitasBaterai = min(100, $this->kapasitasBaterai + $jumlah);
    }

    public function pakaiBaterai(int $jumlah): void {
        if ($jumlah < 0) {
            echo "Jumlah pemakaian tidak valid.\n";
            return;
        }
        $this->kapasitasBaterai = max(0, $this->kapasitasBaterai - $jumlah);
        if ($this->kapasitasBaterai === 0) {
            $this->statusMenyala = false;
        }
    }

    // Invarian 2: tidak bisa menyala jika baterai 0
    public function nyalakan(): void {
        if ($this->kapasitasBaterai <= 0) {
            echo "Gagal menyalakan: baterai habis!\n";
            return;
        }
        $this->statusMenyala = true;
        echo $this->model . " berhasil dinyalakan.\n";
    }

    public function matikan(): void {
        $this->statusMenyala = false;
        echo $this->model . " dimatikan.\n";
    }

    public function getInfo(): void {
        echo "Merk        : {$this->merk}\n";
        echo "Model       : {$this->model}\n";
        echo "Nomor Seri  : {$this->nomorSeri}\n";
        echo "RAM         : {$this->jumlahRAM} GB\n";
        echo "Baterai     : {$this->kapasitasBaterai}%\n";
        echo "Status      : " . ($this->statusMenyala ? "Menyala" : "Mati") . "\n";
    }
}

// Program utama
$laptop = new LaptopAsus("ROG Strix G16", "ASUS-2024-001", 16);
$laptop->getInfo();

// Operasi sah
$laptop->nyalakan();
$laptop->pakaiBaterai(30);
$laptop->getInfo();

// Operasi tidak sah 1: cas melebihi 100
$laptop->cashBaterai(200);
$laptop->getInfo(); // tetap maksimal 100

// Operasi tidak sah 2: nyalakan saat baterai habis
$laptop->pakaiBaterai(100); // baterai jadi 0, otomatis mati
$laptop->nyalakan(); // harus ditolak
$laptop->getInfo();